### PR TITLE
Fix bn layer backwarding @open sesame 09/18 14:15

### PR DESCRIPTION
**Pull request by zhoonit**
**Changes proposed in this PR:**
- Change formula to calculate bn layer backward propagation
- Enable conv2d test case
- Add test case when batch is 1
- Renew test case to use non constant grad_ys

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

File: nntrainer/src/bn_layer.cpp [hunk 1]
```diff
@@ -161,27 +160,30 @@ sharedConstTensor
...
 
-  int batch = input_dim.batch();
+  int N = 1;
+
+  for (auto &axis : axes_to_reduce) {
+    N *= input_dim.getTensorDim(axis);
+  }
 
-  dgamma = x_normalized.multiply(deriv).sum(axes_to_reduce);
+  ///@note faster implementation. Should work but it does not match with tf
```
Comment (zhoonit) [inline nntrainer/src/bn_layer.cpp:175]: remove this line. it matches now

File: nntrainer/include/tensor.h [hunk 2]
```diff
@@ -302,6 +302,13 @@ class Tensor {
    */
   Tensor divide(Tensor const &m) const;
 
+  /**
+   * @brief    Tensor power Element by Element
+   * @param[in] float Divisor Tensor
+   * @retval Calculated Tensor
+   */
+  Tensor pow(float m) const;
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:310]: does it give us better accuracy?

Comment (zhoonit) [reply nntrainer/include/tensor.h:310]: The validation result says it is equivalent to the very small tolerance. I made this for checking purposes but decided leave here as it is a basic operation.

File: nntrainer/src/tensor.cpp [hunk 3]
```diff
@@ -238,6 +239,15 @@ int Tensor::subtract_i(float const &value) { return this->add_i(-value); }
 
 Tensor Tensor::subtract(float const &value) { return this->add(-value); }
 
+Tensor Tensor::pow(float exponent) const {
+  Tensor result(dim);
+
+  for (unsigned int i = 0; i < length(); ++i)
+    *result.getAddress(i) = powf(*getAddress(i), exponent);
```
Comment (kparichay) [inline nntrainer/src/tensor.cpp:246]: use std::transform than a loop here.
Also, i wonder about the accuracy of pow for the exponent of 2 vs using cblas multiply? Is using `powf` more accurate?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 4]
```diff
@@ -853,21 +853,57 @@ TEST_F(nntrainer_BatchNormalizationLayer, forward_backward_training_01_p) {
...
-  EXPECT_NO_THROW(backward_result = *layer.backwarding(
-                    MAKE_SHARED_TENSOR(constant(1.0, 3, 1, 1, 12)), 1));
+  nntrainer::Tensor backward_in(layer.getOutputDimension());
+  loadFile("tc_bn_fc_1_goldenBNLayerBackwardDxIn.out", backward_in);
+
+  nntrainer::Tensor backward_result =
+    *layer.backwarding(MAKE_SHARED_TENSOR(backward_in), 1);
 
-  matchOutput(backward_result, "tc_bn_fc_1_goldenBNLayerBackwardDx.out");
+  // matchOutput(backward_result, "tc_bn_fc_1_goldenBNLayerBackwardDx.out");
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:864]: intentional?

Comment (zhoonit) [reply test/unittest/unittest_nntrainer_layers.cpp:864]: Oh no.. seems like it hasn't verified yet for conv

File: nntrainer/src/tensor.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/src/tensor.cpp:246]: Accuracy is verified with (0.5f and `sqrtFloat`)
This is for negative and non-integer power.

Updated to use apply instead

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 5]
```diff
@@ -853,12 +853,13 @@ TEST_F(nntrainer_BatchNormalizationLayer, forward_backward_training_01_p) {
...
 
   EXPECT_NO_THROW(forward_result = layer.forwarding(MAKE_SHARED_TENSOR(in)));
-
   matchOutput(*forward_result, "tc_bn_fc_1_goldenBNResultForward.out");
 
-  nntrainer::Tensor backward_result;
-  EXPECT_NO_THROW(backward_result = *layer.backwarding(
-                    MAKE_SHARED_TENSOR(constant(1.0, 3, 1, 1, 12)), 1));
+  nntrainer::Tensor backward_in(layer.getOutputDimension());
+  loadFile("tc_bn_fc_1_goldenBNLayerBackwardDxIn.out", backward_in);
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:859]: this test the backwardDx. `AfterUpdate` will be in next unittest?

Comment (zhoonit) [reply test/unittest/unittest_nntrainer_layers.cpp:859]: Yes. Do you prefer it to be deleted now?

Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:859]: No its fine. I just wanted to know.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**